### PR TITLE
Enhanced readability with colors

### PR DESCRIPTION
**Pull request by glehmann**
color the status in yellow or red, depending on the severity

   no color is used for normal status, like Running. Green is not used in
   order to avoid the confusion with red color for color blind users.

   yellow is used for warnings and transional status, like Pending - very
   useful to differentiate Pending and Running.

   red is used for failed, error, ... status

   because strings contain the most information, and white is the most
   readable color, strings are changed to white color.

   table output is changed to two alternating colors, to enhance the
   readability of the columns without using too many colors

   key indent color alternates between yellow and cyan to enhance the indent
   readability and differentiate the key from the content

   while it would be great to enhance the readability of the help page,
   coloring everything in yellow doesn't seem to help for that purpose. As for
   strings, white is more readable and seems more suited for this complex
   content

## Related issue (if exists)

### CODE REVIEW

**Comment by glehmann** [comment]
I'm very new to this project, so I'm not sure if this has already been discussed and maybe rejected. Please just tell me if that's the case!

If you're interested, I'll take a bit more time to fix the tests :)

**Comment by dty1er** [comment]
Hi @glehmann , thank you for much effort!

> color the status in yellow or red, depending on the severity

I totally agree with this concept. Actually this was what I wanted to do.
This PR looks promising, but let me leave some comments here

**Comment by dty1er** [comment]
Sorry, I mistakenly closed the PR 😢 nvm

File: printer/kubectl_output_colored_printer.go [hunk 1]
```diff
@@ -17,6 +17,104 @@ type KubectlOutputColoredPrinter struct {
 	Recursive      bool
 }
 
+func ColorStatus(status string) (color.Color, bool) {
```
Comment (dty1er) [inline printer/kubectl_output_colored_printer.go:20]: This looks awesome. I was too lazy to do this 😅

**Comment by dty1er** [comment]
> yellow is used for warnings and transional status, like Pending - very
useful to differentiate Pending and Running.
> red is used for failed, error, ... status

This sounds good.

**Comment by dty1er** [comment]
> table output is changed to two alternating colors, to enhance the
readability of the columns without using too many colors

I want to discuss this.
I intentionally used many colors in table output because I know some usecases to use kubecolor for kubernetes workshop or something like that. In the workshop, it is easy to tell "see the green column. It means~" with using many colors.
I am not disagreeing using only 2 colors, but can I ask you why you thought using 2 colors is better?

**Comment by dty1er** [comment]
> key indent color alternates between yellow and cyan to enhance the indent
readability and differentiate the key from the content

This basically sounds good.
Now I am thinking if we choose yellow and cyan for the key color, it might better not to use them in "values" color.

**Comment by dty1er** [comment]
> while it would be great to enhance the readability of the help page,
coloring everything in yellow doesn't seem to help for that purpose. As for
strings, white is more readable and seems more suited for this complex
content

Fair enough. In my opinion, there are 2 chances for users to see help when using kubectl:
* When a user specified -h or --help
* When a user made mistake in the command usage

For example, `kubectl top` and `kubectl top -h` show the same "help" content, but the result code is different: `kubectl top` returns 1 but `kubectl top -h` returns 0 (according to `echo $?`).

I think this is because `kubectl top` is simply wrong usage because `top` requires arguments, but `kubectl top -h` is not wrong.
I want to make these 2 cases clear in kubecolor as well. Because `kubecolor top` is wrong usage (result code is 1), the shown help should be colored in yellow. However, `kubecolor top -h` don't need to be shown in yellow; it should be ok to show in white as you suggested.

**Comment by dty1er** [comment]
Currently my concerns are mainly 2: table colors and key-value color duplication. It would be helpful if you can tell me some thoughts.

I am neither UI designer nor color scheme professional, so I was wondering I want to improve kubecolor's looking, but I actually had no idea how to do that. I am now very happy to know your opinion about the color. I strongly want to merge this PR after making things clear.

**Comment by glehmann** [comment]
> > table output is changed to two alternating colors, to enhance the
> > readability of the columns without using too many colors
> 
> I want to discuss this.
> I intentionally used many colors in table output because I know some usecases to use kubecolor for kubernetes workshop or something like that. In the workshop, it is easy to tell "see the green column. It means~" with using many colors.
> I am not disagreeing using only 2 colors, but can I ask you why you thought using 2 colors is better?

for several reasons:

* the main reason is to have quite neutral colors, so its easy to highlight important things in the table, like the `Pending` status, with other colors and have a high contrast between the highlighted and non highlighted text. If there is a lot of colors, it becomes quite difficult to highlight things and make it easily eye catching :)
* unix command line tool are usually not very colored. I use that characteristic to contrast with my shell prompt, which is often quite colored, and easily separate the commands and their output
* based on the previous points, we could say that the text should be mostly white, but I still find easier to read the table when the column or better identified. The alternating colors really help here, especially when the table is larger than the terminal and some lines are wrapping on the line below. It's also easier to identify the columns when the table doesn't fit in the terminal and the column header is not visible. Cyan looks like a good secondary color, as its quite neutral, and still let a lot of room to contrast with red or yellow for example.
* finally, and purely subjectively, I found more pleasing to look at a more sober version :)

In summary, mostly for contrast, and a bit for subjective aesthetic :)

**Comment by glehmann** [comment]
> > while it would be great to enhance the readability of the help page,
> > coloring everything in yellow doesn't seem to help for that purpose. As for
> > strings, white is more readable and seems more suited for this complex
> > content
> 
> Fair enough. In my opinion, there are 2 chances for users to see help when using kubectl:
> 
> * When a user specified -h or --help
> * When a user made mistake in the command usage
> 
> For example, `kubectl top` and `kubectl top -h` show the same "help" content, but the result code is different: `kubectl top` returns 1 but `kubectl top -h` returns 0 (according to `echo $?`).
> 
> I think this is because `kubectl top` is simply wrong usage because `top` requires arguments, but `kubectl top -h` is not wrong.
> I want to make these 2 cases clear in kubecolor as well. Because `kubecolor top` is wrong usage (result code is 1), the shown help should be colored in yellow. However, `kubecolor top -h` don't need to be shown in yellow; it should be ok to show in white as you suggested.



> > while it would be great to enhance the readability of the help page,
> > coloring everything in yellow doesn't seem to help for that purpose. As for
> > strings, white is more readable and seems more suited for this complex
> > content
> 
> Fair enough. In my opinion, there are 2 chances for users to see help when using kubectl:
> 
> * When a user specified -h or --help
> * When a user made mistake in the command usage
> 
> For example, `kubectl top` and `kubectl top -h` show the same "help" content, but the result code is different: `kubectl top` returns 1 but `kubectl top -h` returns 0 (according to `echo $?`).
> 
> I think this is because `kubectl top` is simply wrong usage because `top` requires arguments, but `kubectl top -h` is not wrong.
> I want to make these 2 cases clear in kubecolor as well. Because `kubecolor top` is wrong usage (result code is 1), the shown help should be colored in yellow. However, `kubecolor top -h` don't need to be shown in yellow; it should be ok to show in white as you suggested.

seems good to me :+1:

**Comment by glehmann** [comment]
> > key indent color alternates between yellow and cyan to enhance the indent
> > readability and differentiate the key from the content
> 
> This basically sounds good.
> Now I am thinking if we choose yellow and cyan for the key color, it might better not to use them in "values" color.

It would be better, in order to have a good contrast between key in values, but is it possible with the few available colors?

**Comment by glehmann** [comment]
> Currently my concerns are mainly 2: table colors and key-value color duplication. It would be helpful if you can tell me some thoughts.
> 
> I am neither UI designer nor color scheme professional, so I was wondering I want to improve kubecolor's looking, but I actually had no idea how to do that. I am now very happy to know your opinion about the color. I strongly want to merge this PR after making things clear.

I'm not an UI designer either, but I had the chance to work during several years on a command line tool that strongly focused on usability. We tested many options, and that allowed me to build a quite strong opinion on how colors (and cli in general) should be used :)

Thanks for creating this nice tool and to be open to discuss about it :)

**Comment by dty1er** [comment]
@glehmann Sorry for my very late response! I was getting busy because just some work suddenly came. Let me reply to your comments...

**Comment by dty1er** [comment]
> In summary, mostly for contrast, and a bit for subjective aesthetic :)

Thank you for your clarifications.
The first point makes sense to me. I actually wanted to highlight the important things (e.g. "CrashLoopBackoff") in kubecolor, but using many colors makes it difficult. I still find some benefits on using many colors, but it looks like a tradeoff, and now I think using only 2 colors will be better.

**Comment by dty1er** [comment]
> It would be better, in order to have a good contrast between key in values, but is it possible with the few available colors?

Currently kubecolor uses [some colors](https://github.com/dty1er/kubecolor/blob/main/printer/colors_definition.go#L28-L34):

* key colors
  * Now we are going to use cyan and yellow for this
* value colors
  * string
  * number
  * bool 
  * null (including "none" and "unknown")

Usually, most "value" is String type. bool and null rarely appears, so I think string should be colored neither in cyan nor yellow. (should green look better for example?)
Fortunately we still have magenta and white and red, so technically we can use them for number, bool, and null. However, red might not be good because it might look showing some "error".

**Comment by rothgar** [comment]
I appreciate the additional color indicators for errors. If possible avoiding reds and greens that are the same shades would be helpful as I am red/green color blind. Usually if a darker green is used but a brighter/light red it's easier for me to tell the difference.

**Comment by glehmann** [comment]
sorry @dty1er I somehow missed your comments. Here are a few snapshots to illustrate the output of kubecolor with my changes:
kubecolor
kubecolor1
kubecolor2

strings output look good to me in white, but other color would work also :)

@rothgar what do you think of the colors in the status of the first image? Are they easy enough to read? There is no green, so I hope they are easy enough to read :)

**Comment by rothgar** [comment]
I can easily differentiate the colors in that screenshot so they LGTM

**Comment by dty1er** [comment]
related: https://github.com/dty1er/kubecolor/discussions/61

**Comment by rkennedy-tpl** [comment]
> > yellow is used for warnings and transional status, like Pending - very
> > useful to differentiate Pending and Running.
> > red is used for failed, error, ... status
> 
> This sounds good.

this feature is exactly why I came looking for this tool. when looking at a long list of `Pending` vs `Running` containers, the same character length and similar shapes blur together easily. if I'm gonna use a colorizing tool, I want it to go kinda "all-in". I don't find the rainbow of text overwhelming, I find that color provides a "second axis" of information I can consume simultaneously and/or faster. all these screenshots are wonderful to me => https://github.com/dty1er/kubecolor/pull/40#issuecomment-813382179

**Comment by prune998** [comment]
Any chance this is going to be merged ?

**Comment by bruyneron** [comment]
This is really useful. What's holding back from merging this?

**Comment by prune998** [comment]
@glehmann could you rebase your PR on the latest main branch ? That will re-trigger the tests... maybe it will help the owner to merge this branch ?

**Comment by hidetatz** [comment]
closing now as it's stale.
